client/state/data-layer/extensions-middleware.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * External dependencies
+ */
+import { toPairs, fromPairs } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { local, mergeHandlers } from './utils';
+
+const stripHandlersForKey = ( key, keyHandlers, keyHandlersToStrip ) => {
+	const newKeyHandlers = (
+		keyHandlersToStrip
+			? keyHandlers.filter( ( handler ) => ! keyHandlersToStrip.includes( handler ) )
+			: keyHandlers
+	);
+
+	return [ key, newKeyHandlers ];
+};
+
+const stripHandlers = ( handlers, handlersToStrip ) => {
+	return fromPairs(
+		toPairs( handlers ).map(
+			( [ key, keyHandlers ] ) => stripHandlersForKey( key, keyHandlers, handlersToStrip[ key ] )
+		)
+	);
+};
+
+export const extensionsHandlers = () => {
+	let handlers = {};
+
+	/**
+	 * Dynamically add handlers to this middleware.
+	 *
+	 * @function
+	 * @param {Object} handlersToAdd The actions to be added.
+	 * @return {Function} removeHandlers, removes the handlers that were added.
+	 */
+	const addHandlers = ( handlersToAdd ) => {
+		handlers = mergeHandlers( handlers, handlersToAdd );
+
+		return function removeHandlers() {
+			handlers = stripHandlers( handlers, handlersToAdd );
+		};
```

Comment:
what if here we autogenerated a name/key for the handler and used that key to simplify insertion and removal. I don't think it would even be that much of a pain point to completely rebuild the tree on each of such operation vs actually removing the handlers from the existing bundle.

rebuilding on each change could be conceptually simpler and easier to debug as well